### PR TITLE
Fix #91958: use `mach_timebase_info` to determine process time coefficient on macOS

### LINKED ISSUE #91958: Process::TotalProcessorTime is about 42 times lower than expected on ARM64 Mac
### Description

Let's consider this process on my system that has PID of 66358 (as shown by the Activity Monitor):
image

It has about 22 hours of CPU time.

### Reproduction Steps

Write and run this program:
```csharp
using System.Diagnostics;

var process = Process.GetProcessById(66358);
Console.WriteLine(process.TotalProcessorTime);
Console.WriteLine(process.TotalProcessorTime * 42);
```

### Expected behavior

This program should output about 22 hours in the first line, and some nonsensical value in the second line.

### Actual behavior

```
00:00:32.3739259
00:22:39.7048878
```

### Regression?

_No response_

### Known Workarounds

Multiply the result by 42 on ARM64 Mac :)

### Configuration

Macbook Pro M2 Max, ARM64
.NET SDK 7.0.202

### Other information

This is not the first time this API behaves problematically on macOS. The previous fix was https://github.com/dotnet/corefx/pull/37637

I suspect that the fix was incorrect: in reality, we should call `mach_timebase_info` to get the coefficient to convert the value returned by the underlying API to nanoseconds, and apparently that value may be system-dependent.

### PR DESCRIPTION
**Pull request by ForNeVeR**
Closes #91958.

I have compared the results of the process timing functions to the results of `ps` on my M2 MacBook device, and they seem to work properly after the changes.

The new tests were properly failing before the change (since we were returning values 42 times lower than the native results), and are, of course, green after the changes.

### CODE REVIEW

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs
Comment (ForNeVeR) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:120]: Please check if this is correct.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs
Comment (ForNeVeR) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:1053]: Please help me to figure out what to do with this.

Apparently, the `Interop.*.cs` files I've added to the test assembly relies on resources/localization, but I haven't figured out how to reuse the main projects' resources in tests.

I think it's possible to just duplicate the API I use in tests here and get rid of `SR`. But this will be quite a lot of duplication, since the structure returned by `proc_pid_rusage` is kinda long.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs
Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:120]: One possibility I see here would be to just return a coefficient of `1` in case of an error in this place, to fall back to the old behavior.

Comment (tmds) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:120]: > One possibility I see here would be to just return an effective coefficient of numer: 1, denom: 1 in case of an error in this place, to fall back to the old behavior.

Yes, let's do this.

And, let's add a `Debug.Assert` so debug test runs will fail if this would ever return non-success.

Should we account for `mach_timebase_info` not existing on some systems?

Comment (am11) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:120]: I think our minimum macOS support offsets this API’s minimum requirement.

Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:120]: Apparently [it was available at least in OS X 10.10.1](https://stackoverflow.com/a/31454136/2684760) back in 2015, and .NET only supports 10.15+, so we should be okay.

I'll apply the suggestion.

Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: Done!

File: src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs
Comment (adamsitnik) [inline src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs:16]: @jkoritzinsky @AaronRobinsonMSFT In terms of marshaller best practices, do we need to explicitly specify sequential layout for such structs?

```suggestion
        [StructLayout(LayoutKind.Sequential)]
        public struct mach_timebase_info_data_t
```

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:13]: It would be better to initialize this field lazily, when we need it for the first time. Otherwise this sys-call:
- may be called even if we don't need it (during type initializaiton)
- in theory it may fail and be very hard to handle properly

Since we use the following logic in 4 places:

```cs
return new TimeSpan(Convert.ToInt64($ulong * timeBase.numer / timeBase.denom / NanosecondsTo100NanosecondsFactor));
```

 we could introduce a helper method that could take care of everything, something like this:

```suggestion
        private static volatile uint s_timeBase_numer, s_timeBase_denom;
        
        private static TimeSpan Map(ulong sysTime)
        {
            uint denom = s_timeBase_denom;
            if (denom == default)
            {
                Interop.libSystem.mach_timebase_info_data_t timeBase = GetTimeBase();
                s_timeBase_denom = denom = timeBase.denom;
                s_timeBase_numer = timeBase.numer;
            }
            uint numer = s_timeBase_numer;
            
            return new TimeSpan(Convert.ToInt64(sysTime * numer / denom / NanosecondsTo100NanosecondsFactor));
        }
```

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:86]: thank you for finding and updating all use cases! 👍

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs
Comment (adamsitnik) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:1053]: That is a great question!

You should be able to reference the resource file in the test project similar to:

https://github.com/dotnet/runtime/blob/913231acc359c5e94d4397e8fe0c8a38ae6b0c5a/src/libraries/System.Formats.Tar/tests/System.Formats.Tar.Tests.csproj#L5

**Review by adamsitnik — COMMENTED** [review]
Overall it looks good, but I found few places that could be polished a bit.

Thank you for your contribution @ForNeVeR !

File: src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs:16]: Technically, no. Value types in .NET default to sequential layout. However, and this is annoying, there are some Roslyn warnings that are suppressed if one does explicitly mark the type with sequential layout. The reasoning here is historical, but the gist is if Roslyn complains about unreferenced fields, which can happen for types used in interop, then placing `StructLayout(LayoutKind.Sequential)` on the type will automatically suppress the warning.

The interop team's general guidance here has been to accept the defaults except where there is annoying friction with C# or where the tooling requires explicit details. This falls into the C# friction bucket, but only if a warning is emitted.

Comment (adamsitnik) [reply src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs:16]: @AaronRobinsonMSFT thank you for a very detailed answer!

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs
Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: For some reason, this only started working for me after I added OSX-specific TFM to the test project. I guess it has something to do with how the resource layout works? Anyway, please take another look at the current solution. I have to say I am a bit confusing around this part.

File: src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs
Comment (ForNeVeR) [reply src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs:16]: I can't see any related warnings in the compilation (and also we seem to have warnings as errors enabled in this part?).

Does this mean this attribute is unnecessary? I am totally okay with adding that if required. Though, yeah, we all know that sequential is the default struct layout 😅

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs
Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: Yeah, agree with your reasoning and applied a bit modified version of this snippet (the only change is in naming, I like `MapTime` a little bit better).

File: src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/Common/src/Interop/OSX/Interop.libSystem.cs:16]: > Does this mean this attribute is unnecessary?

Yes.

**Comment by ForNeVeR** [comment]
Build on CI is now failing due to something being wrong with `SR` on Windows. Have I broken that? I need help with resources, I don't think my changes to `StringResourcesPath` were right.

File: src/libraries/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests.csproj
Comment (am11) [inline src/libraries/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests.csproj:8]: We have two options:

1. Include the two keys in test resources:
    ```xml
	  <data name="CantGetAllPids" xml:space="preserve">
	    <value>Could not get all running Process IDs.</value>
	  </data>
	  <data name="RUsageFailure" xml:space="preserve">
	    <value>Failed to set or retrieve rusage information. See the error code for OS-specific error information.</value>
	  </data>
    ```
    in:
    https://github.com/dotnet/runtime/blob/b4a3c9e6fbe09d2f52465e2189d57ef5c04b94c3/src/libraries/System.Diagnostics.Process/tests/Resources/Strings.resx#L1
    and delete this hard-coded line.

2. Include test resources alongside source ones -- separated by semicolon `;`:
    ```xml
		 <StringResourcesPath>$(MSBuildProjectDirectory)\Resources\Strings.resx;$(MSBuildProjectDirectory)\..\src\Resources\Strings.resx</StringResourcesPath>
    ```

    multiple resources seem to be supported.

Option 1 is probably much cleaner that avoids mixing stuff, but I'll defer to others. cc @ViktorHofer

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs
Comment (danmoseley) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:125]: Should it do the last division first, like here?

https://chromium.googlesource.com/chromium/src/+/refs/tags/58.0.3029.141/base/time/time_mac.cc#43

(I don't know what values it typically returns)

Comment (danmoseley) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:125]: Oh, this says it gets the 42 from 125/3. Does 125 * ticks go over ulong max?

https://stackoverflow.com/questions/69318594/how-to-get-the-true-value-of-the-system-clock-in-raw-ticks-i-e-mach-absolute-t

Comment (danmoseley) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:125]: Also I suppose you can store 1000 * denom 

I'm sure someone will point out that won't necessarily give the exact same result @tannergooding 🙂

File: src/libraries/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests.csproj
Comment (ViktorHofer) [reply src/libraries/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests.csproj:8]: Agreed. Option 1 is cleaner while it adds some duplication.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs
Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:125]: That's a good point, but I was concerned about losing some precision due to doing division first. Though I can't wrap my head around it right now. Will try to do more thorough analysis later today.

Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:125]: Ok, here are my thoughts.

Typical tick values are pretty large actually. For comparison, let's consider a program that was working on a single-core CPU for a year.
```
> [TimeSpan]::FromDays(365).Ticks
315360000000000
> [long]::MaxValue
9223372036854775807
```

(here I consider `long` since `TimeSpan` takes `long` in its ctor, ot `ulong`)

This means we will overflow on a `numer` of `29248` (or if our tick value will get significantly bigger, i.e. we take 29000 years into account, or a CPU with 29k cores).

This is very far from being realistic, but it is also much closer than I expected. So, to be on the safe side, let's do the same as Chromium does: divide by our factor first, to get a hundred times more space, by losing some precision.

We may, of course, also do `value / (denom * 100) * numer` which is relatively the same, yet will lose a bit more precision as well.

Comment (danmoseley) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: I'm also OK with asserting that numer is reasonably small ( say < 1000) as an alternative, since we suspect that will always be true.

Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: My current suggestion is to leave this as `sysTime / NanosecondsTo100NanosecondsFactor * numer / denom`.

I am not sure how useful an assertion would be in this code. Perhaps it'd be better to add `checked`, since we are concerned by overflow? We do not expect this code to be too performance sensitive, so `checked` might be good.

What do you think?

**Comment by ForNeVeR** [comment]
I believe that all the existing feedback is addressed, so I'm waiting for further feedback. Thank you so much for your time, folks.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs
Comment (danmoseley) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: Not my area, but I"m not sure an exception would be an improvement. I'd be interested in thoughts of a numeric expert like @tannergooding . I'll step aside for area owners to sign off overall...

Comment (ForNeVeR) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs]: I would say that an explicit exception might be an improvement (especially since it would be thrown from a member property such as `TotalProcessorTime` and not on class init).

But of course, I am ready to listen to other suggestions.

**Comment by marek-safar** [comment]
@dotnet/area-system-diagnostics-process please review

**Comment by EgorBo** [comment]
Ping @dotnet/area-system-diagnostics-process

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you for your fix @ForNeVeR !

And apologies for the delay (I was on a parental leave).

File: src/libraries/System.Diagnostics.Process/tests/Resources/Strings.resx
Comment (adamsitnik) [reply src/libraries/System.Diagnostics.Process/tests/Resources/Strings.resx:123]: https://github.com/dotnet/runtime/pull/93539

**Comment by ForNeVeR** [comment]
Thank you so much for help! ❤️

**Comment by jeffhandley** [comment]
Backporting this fix to 8.0 since this was reported again in #98121.

**Pull request merged**
